SPI: STM32: Allow 16bit sends in 8bit mode if possible in master-only mode.

psidhu opened the pull request:
For example, if we're receiving 24bits, we can receive 16bits in one go, then 8bits in the next saving time. I found this to save, on average, 2.5us on my system.

Tested on nucleo_l4r5zi.

Sorry for the ugly git diff that is produced. I can try to fixup the commit a bit to help if anyone wants.

@erwango Regarding the `ll_func_set_fifo_threshold_16bit` function, is the MP1 setting correct? I only took a guess at it and haven't tested it.

psidhu committed:
spi: stm32: Add ll_func_set_fifo_threshold_16bit helper

Add this function to allow the fifo to be set for 16bits.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

tbursztyka commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -73,71 +73,123 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
 
-	return tx_frame;
-}
+#if defined(CONFIG_SPI_STM32_HAS_FIFO)
```
you could use IS_ENABLED() instead to avoid #ifdef

erwango replied:
If I'm correct, this won't work as `ll_func_set_fifo_threshold_xbit` functions are not defined if CONFIG_SPI_STM32_HAS_FIFO not defined.

tbursztyka replied:
indeed, I missed the fact the new inline function was within a former ifdef.

psidhu commented:
I missed adding the case for STM32MP1 for starting a master transfer. Added that.

erwango commented:
@psidhu 

> Sorry for the ugly git diff that is produced. I can try to fixup the commit a bit to help if anyone wants.

Indeed, it would be good. 
Though before doing this:
Following the modifications you're bringing, maybe some re-factorization would be welcome. Indeed, this change modifies usefulness of some functions. So maybe there is some more work to do to make things a bit nicer. For instance:
`spi_stm32_shift_m` is nearly empty now (what about doing more in `spi_stm32_shift_frames` and removing `spi_stm32_shift_m`.
`spi_stm32_next_tx` is useless (factorization function called only once).

Can you have a check?

psidhu commented:
@erwango please recheck.

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -73,28 +73,41 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
-			tx_frame = UNALIGNED_GET((u16_t *)(data->ctx.tx_buf));
-		}
+	u16_t rx_frame;
+	u8_t frame_sz;
```
Since only use of this variable it to indicate if we proceed with 16bits transfer or not, 
for code reading purpose,  it would be easier to set it as a bool with a meaningful name

psidhu replied:
Ok.

erwango reviewed (changes requested):
Thanks for this rework, this is much easier both to review and to read.
I have a last comment on a variable name and a last request:
If not too much work, can you split out the last commit in 2 ?
- first with removal
- second with the main change
It would make change quite straightforward.

psidhu commented:
> If not too much work, can you split out the last commit in 2 ?
> 
>     * first with removal
> 
>     * second with the main change
>       It would make change quite straightforward.

@erwango Seems easy enough. I'll have something pushed shortly.

psidhu committed:
spi: stm32: remove spi_stm32_next_tx helper

Remove usage of this helper to stage for 16bit frame transfers in 8bit
mode.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

psidhu committed:
spi: stm32: Enable sending 16bits at a time in 8bit mode if possible

If possible, allow sending 16bits at once instead of 8bits. I found
large delays (up to 3us) between sending bytes due to Zephyr SPI
overhead, so allowing 16bits at a time if possible helps save that
time.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

psidhu commented:
@erwango please recheck.

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -73,28 +73,14 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
 
-	tx_frame = spi_stm32_next_tx(data);
+	tx_frame = (spi_context_tx_on(&data->ctx)) ?
+		UNALIGNED_GET((u8_t *)(data->ctx.tx_buf)) : SPI_STM32_TX_NOP;
```
Are we sure `(SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8` is true here?

psidhu replied:
Oh, I'm sorry. I introduced that when splitting out the commits. Though not important, I'll fix it.

erwango reviewed (commented):
Last point otherwise I'm fine with the change

psidhu commented:
@erwango Please recheck.
